Azure Monitor: E2E Tests

aangelisc opened the pull request:
This PR begins building out the E2E testing suite for Azure Monitor.

The following E2E tests have been added:

- Add datasource and import a dashboard
- Add datasource and create a metrics query
- Add datasource and create a logs query
- Add datasource and create an ARG query
- Add datasource with multiple template variables and ensure query functions (also tests advanced resource picker configuration). The following Azure Monitor template queries are covered:
    - Subscription
    - Resource Groups
    - Namespaces
    - Resources
- Add datasource with Azure Monitor annotation

Additional tests that can be added in future:

- Explore page test
- Metrics query test with dimensions
- Edit a panel
- Aliases
- Query history for log analytics and ARG queries
- Template variables covering:
    - Metric names
    - Workspaces
    - Logs
    - ARG
- Test link to Azure portal

Some caveats before this can be merged:

- Need to confirm the process for adding these E2E tests to drone.
- Need to identify how to obscure values in Cypress videos.
- Need to change configuration in Azure for consistency.

Other notes:
- The `configurePanel` flow had to be updated because Cypress couldn't find the `RefreshPicker` element (due to a `display: none` CSS property somewhere).
- The `addDashboard` flow can be updated to support providing a function for `datasourceForm` to the `addVariables` function (similar to the `addAnnotations` flow). This will simplify the template variable test as currently the whole flow is manually written.
- Betterer results had to be updated as the `e2e` object has properties with `any` assertions (`env, getScenarioContext, setScenarioContext`).

Fixes #53042

aangelisc committed:
Update credentials form with data-testids and aria-labels

aangelisc committed:
Update gitignore

aangelisc committed:
Add example dashboard

aangelisc committed:
Stub out E2E test for creating ds and importing dashboard

aangelisc committed:
Add component selectors

aangelisc committed:
Remove subscription check temporarily

aangelisc commented:
A simple E2E test that adds the datasource and imports an example Azure Monitor dashboard. 

TODO:

- [ ] Component selectors don't seem to be working as expected so currently using `aria-label` which is not the preferred method.
- [ ] Assert on chosen subscription value.

aangelisc committed:
Appropriately set disabled prop

aangelisc committed:
Fix lint issues

aangelisc committed:
Update to use selectors and wait on subscriptions request

aangelisc committed:
Add test for metrics panel

- Add required selectors for resource picker and metrics query editor

aangelisc committed:
Add logs and ARG basic query scenarios

- More selector updates

aangelisc committed:
Add E2E test for template variables

- Tests advanced resource picker
- Adds required selectors

aangelisc committed:
Remove log and add annotation e2e test

aangelisc committed:
Update test

aangelisc committed:
Merge branch 'main' into azure-monitor-e2e

aangelisc committed:
Prettier/betterer updates

- Remove gitignore change

aangelisc committed:
Lint issues

aangelisc committed:
Update betterer results

aangelisc committed:
Lint issue and remove unneeded import

aangelisc committed:
Don't print certain commands

aangelisc committed:
Merge branch 'main' into azure-monitor-e2e

aangelisc committed:
Avoiding flakiness

- Ensure code editor has sufficient time to load in ARG test
- Avoid flakiness around correct template variable being selected by typing in resource name

aangelisc committed:
Remove be.visible requirement

aangelisc committed:
Merge branch 'main' into azure-monitor-e2e

aangelisc committed:
Update test

andresmgot reviewed (commented):
Awesome work here! Really nice coverage for a first version. It's a pity we cannot run it on drone _yet_. I'd assume that we are going to see errors due to the environment but we can fix those later.

I tried to run it, with a docker container running 9.2.0-beta1 and running `yarn e2e:dev`  but I am getting this error:

![Screenshot from 2022-10-06 17-23-09](https://user-images.githubusercontent.com/4025665/194353711-f6b4f4c8-3d17-47e8-8f4b-038b96042065.png)

Which is weird, because I can browse grafana normally in a real browser. How are you running the tests?

aangelisc commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -0,0 +1,310 @@
+function provisionAzureMonitorDatasources(datasources: AzureMonitorProvision[]) {
+  const datasource = datasources[0].datasources[0];
+
+  e2e()
+    .intercept(/subscriptions/)
+    .as('subscriptions');
```
This allows us to intercept the subscriptions API request so that we can wait for the response on line 42 before continuing. I was previously having issues with the response being flaky in response time.

aangelisc commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -0,0 +1,310 @@
+            e2eSelectors.queryEditor.resourcePicker.apply.button().click();
+            e2e().contains('microsoft.storage/storageaccounts');
+            e2eSelectors.queryEditor.metricsQueryEditor.metricName.input().find('input').type('Used capacity{enter}');
+          },
```
Currently the E2E flow for `addPanel` ensures that `chartData` is called and returns correctly. I could try to think of a visual element that would be consistent to check against.

andresmgot replied:
ah, right, that's fine then

andresmgot commented:
It seems that there was a transient issue on my side. I got past that original issue and executed _some_ tests but eventually my chrome crashes. Maybe it's due to the length of the test suite and the big resolution set:

![Screenshot from 2022-10-07 13-15-13](https://user-images.githubusercontent.com/4025665/194541109-3cdabdc3-7caf-4c0c-9e16-b863ab0e49e0.png)

It can be something related to my machine or Chrome. Are you able to run these tests @asimpson @yaelleC?

In any case, I think we can simplify (or try to make tests faster) since it takes several minutes to complete these tests. 

![Screenshot from 2022-10-07 13-24-37](https://user-images.githubusercontent.com/4025665/194542717-0e60d4f7-b7e3-477d-953c-97caffa54031.png)

The first one "import dashboard" is not providing much value. For me it's even using the wrong data source. We can probably skip it:

![Screenshot from 2022-10-07 13-01-20](https://user-images.githubusercontent.com/4025665/194541903-80e86bcf-b701-4b7e-bdd4-7dd8b86ebc8b.png)

The three tests that creates a query for metrics, logs and arg can probably be merged into one. We lose the granularity of the different tests but we only need to create one data source for the three of them.

I never got to run the last one, the one for annotations, since my chrome crashes before that but it's probably not critical enough to be run for all the builds. Is there anything specific that you are looking to test with that?

aangelisc commented:
> In any case, I think we can simplify (or try to make tests faster) since it takes several minutes to complete these tests.

The major improvement to speed we could make is to only create a single datasource and then use that datasource for all subsequent tests. That should shave off a fairly substantial chunk of time.

> The first one "import dashboard" is not providing much value. For me it's even using the wrong data source. We can probably skip it:

We could just import one of the standard Azure dashboards here as validation rather than a generic test dashboard.

> The three tests that creates a query for metrics, logs and arg can probably be merged into one. We lose the granularity of the different tests but we only need to create one data source for the three of them.

This would be solved with my first suggestion. Although we could just migrate this to one test and create a panel for each query rather than a separate dashboard for each.

> I never got to run the last one, the one for annotations, since my chrome crashes before that but it's probably not critical enough to be run for all the builds. Is there anything specific that you are looking to test with that?

Similar to testing the template variables I'm just ensuring that an annotation that uses an Azure Monitor query functions as expected.

andresmgot commented:
> The major improvement to speed we could make is to only create a single datasource and then use that datasource for all subsequent tests. That should shave off a fairly substantial chunk of time.
> 

:+1: 

> We could just import one of the standard Azure dashboards here as validation rather than a generic test dashboard.

If I understand that test, it will only fail if the import process fail (which is not something we own in the plugin). If the panels themselves fail, the test will still succeed so there is not much value for us to do that test.

> Although we could just migrate this to one test and create a panel for each query rather than a separate dashboard for each.

That sounds faster :+1: 

> Similar to testing the template variables I'm just ensuring that an annotation that uses an Azure Monitor query functions as expected.

Yeah, I get that, the thing is that the probabilty of a change breaking only annotations is pretty low (I've never seen one) so I wonder if it's worth it the extra time that the CI will take to run it for all PRs :thinking:

aangelisc committed:
Update selector name

yaelleC commented:
I can run the test with `yarn run e2e:dev` without spinning any additional docker container. But somehow I get stuck on `
[data-testid="data-testid azure-cloud"]` step every time the ds is set up even though the `Azure Cloud` field is visible 🤔

aangelisc commented:
> I can run the test with `yarn run e2e:dev` without spinning any additional docker container. But somehow I get stuck on ` [data-testid="data-testid azure-cloud"]` step every time the ds is set up even though the `Azure Cloud` field is visible 🤔

That's interesting, what's the failure? Just that it can't find the element?

aangelisc committed:
Reuse datasource

yaelleC commented:
> That's interesting, what's the failure? Just that it can't find the element?

Yes, I get `Timed out retrying after 30000ms: Expected to find element: [data-testid="data-testid azure-cloud"], but never found it.`

That's what I get when I inspect:
<img width="165" alt="Screenshot 2022-10-10 at 13 50 03" src="https://user-images.githubusercontent.com/42030685/194860108-f54da417-b55c-4d5d-9653-e100da48e548.png">

andresmgot commented:
> I can run the test with `yarn run e2e:dev` without spinning any additional docker container.

I couldn't run them using docker because you need the changes in the functional code that there are in this PR so I wonder maybe you don't have the new data-testid because you have not rebuilt the Grafana frontend?

aangelisc committed:
Fix datasource reuse and combine query tests

aangelisc committed:
Remove import dashboard step as unneeded

andresmgot commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -21,6 +21,8 @@ type AzureMonitorConfig = {
 
 type AzureMonitorProvision = { datasources: AzureMonitorConfig[] };
 
+const dataSourceName = 'Azure Monitor E2E Tests';
```
the problem of using a fixed data source name is that if the tests are stopped in the middle, it won't work after restarting (since it won't delete it). Can you add some randomized string at the end?

yaelleC replied:
I believe if you don't specify the name line 35 the e2e test will come up with a randomized name for you

yaelleC replied:
🤔 I see you use it in `addPanel`. Could be get the name from the datasource created?
If not, adding a randomised string at the end is a good way forward too

aangelisc replied:
I'll add a randomised string at the end. I did originally try to retrieve the random name generated but because of the way the Cypress `wrap` function works this didn't seem like it was possible. The E2E scenarios are structured in such a way that all changes in a scenario are reverted at the end of the scenario which is why I went with the fixed name 😊 

I will also add the `skipScenario` flag to the annotations test and then this should be good to merge!

andresmgot reviewed (approved):
Now I could run the whole test suite, great! :+1: 

I think I'd still leave out the annotation test (maybe just skip it, leaving the code) until we have this running on CI and it's stable but up to you, what you think is best

andresmgot commented:
> > That's interesting, what's the failure? Just that it can't find the element?
> 
> Yes, I get `Timed out retrying after 30000ms: Expected to find element: [data-testid="data-testid azure-cloud"], but never found it.`
> 
> That's what I get when I inspect: <img alt="Screenshot 2022-10-10 at 13 50 03" width="165" src="https://user-images.githubusercontent.com/42030685/194860108-f54da417-b55c-4d5d-9653-e100da48e548.png">

BTW, I also hit this now @yaelleC. Apparently, `yarn:e2e` spins its own Grafana server (running in :3001 for me rather than :3000). In my sytem, the server running in :3000 was running the latest changes but the one in :3001 was not (not sure which build it was using). After killing my server and restarting the `yarn:e2e` command, I could successfully run the tests :shrug: .

aangelisc committed:
Review

- Randomise datasource name
- Skip annotations test
- Remove unused example dashboard

andresmgot reviewed (approved):
:+1:

asimpson reviewed (approved):
Really good work on this! e2e tests are _tricky_ things :trophy:

aangelisc committed:
Update to ensure e2e test works in CI

aangelisc committed:
Update e2e test

- Update environment variables (process is not available in cypress)
- Add wait on resource picker searches to avoid flakiness
- Update subscription and resource group names

aangelisc committed:
Update CODEOWNERS

aangelisc committed:
Parse credentials in CI from outputs file

andresmgot reviewed (approved):
:+1: :+1:

aangelisc committed:
Update outputs file path

aangelisc committed:
Merge branch 'main' into azure-monitor-e2e

aangelisc committed:
Fix selector

aangelisc committed:
Undo selector change

aangelisc committed:
Update e2e tests

- Set default subscription
- Fix datasource selection in variable editor
- Fix resource picker search flakiness
- Set subscription in ARG query test
- Fix resource group selection
- Update resource group

joshhunt commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -0,0 +1,297 @@
+      e2e()
+        .readFile(provisioningPath)
+        .then((azMonitorProvision: string) => {
+          const yaml = load(azMonitorProvision) as AzureMonitorProvision;
```
can we use an eslint-ignore comment instead of adding an exception to betterer?

The plan is for us to remove betterer completely once we fix all the existing lint violations, but we can't do that if we keep adding new stuff to it!

joshhunt replied:
The issue is using `as AzureMonitorProvision` right here. That's where the comment should be added not anything in the e2e package

aangelisc replied:
You're correct, I was confused by the `any` returned in the Grafana E2E package, I've updated this and it should pass.

aangelisc committed:
Review

joshhunt commented on packages/grafana-e2e/src/index.ts:
```diff
@@ -13,6 +13,7 @@ import { getScenarioContext, setScenarioContext } from './support/scenarioContex
 import * as typings from './typings';
 
 const e2eObject = {
+  // eslint-ignore-next-line
```
what rule are we ignoring here? betterer shouldn't be concerned about these lines, and it doesn't look like you changed anything else related to them?

aangelisc replied:
This has been reverted 😊

aangelisc committed:
Review 2
